File: backend/app/api/visits.py
from __future__ import annotations

from datetime import date, datetime, time
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, status

from app.api.deps import AuthenticatedDoctor, get_current_doctor
from app.models.dto import (
  PatientSummary,
  TodayVisitsResponse,
  VisitResponse,
  VisitStatusUpdateRequest,
  VisitUpdateRequest,
  VisitWithPatientResponse,
)
from app.services import visits_service

router = APIRouter(prefix="/visits", tags=["visits"])

CurrentDoctor = Annotated[AuthenticatedDoctor, Depends(get_current_doctor)]


def _map_visit_with_patient(row: dict) -> VisitWithPatientResponse:
  """Map a visit row with joined patient data to response model."""
  patient_data = row.pop("patients", None)
  patient = None
  if patient_data:
    patient = PatientSummary(
      id=patient_data.get("id", ""),
      first_name=patient_data.get("first_name", ""),
      last_name=patient_data.get("last_name", ""),
      phone=patient_data.get("phone"),
      telegram_user_id=patient_data.get("telegram_user_id"),
      telegram_username=patient_data.get("telegram_username"),
      whatsapp_phone=patient_data.get("whatsapp_phone"),
    )
  
  return VisitWithPatientResponse(
    id=row.get("id", ""),
    doctor_id=row.get("doctor_id"),
    patient_id=row.get("patient_id", ""),
    visit_date=row.get("visit_date"),
    visit_time=row.get("visit_time"),
    next_visit_date=row.get("next_visit_date"),
    notes=row.get("notes"),
    medications=row.get("medications"),
    status=row.get("status", "scheduled"),
    status_changed_at=row.get("status_changed_at"),
    status_note=row.get("status_note"),
    rescheduled_to=row.get("rescheduled_to"),
    rescheduled_time=row.get("rescheduled_time"),
    reminder_status=row.get("reminder_status"),
    reminder_sent_at=row.get("reminder_sent_at"),
    reminder_channel=row.get("reminder_channel"),
    created_at=row.get("created_at"),
    patient=patient,
  )


@router.get(
  "/today",
  response_model=TodayVisitsResponse,
  status_code=status.HTTP_200_OK,
)
async def get_today_visits(
  current_doctor: CurrentDoctor,
) -> TodayVisitsResponse:
  """
  Get all visits for today for the current doctor.
  Returns visits ordered by time with patient data included.
  """
  today = date.today()
  rows = visits_service.list_by_doctor_and_date(current_doctor.doctor_id, today)
  
  visits = [_map_visit_with_patient(row) for row in rows]
  
  return TodayVisitsResponse(
    date=today,
    count=len(visits),
    visits=visits,
  )


@router.get(
  "",
  response_model=TodayVisitsResponse,
  status_code=status.HTTP_200_OK,
)
async def get_visits_by_date(
  current_doctor: CurrentDoctor,
  date_param: date = Query(alias="date", default=None, description="Date to get visits for (YYYY-MM-DD)"),
) -> TodayVisitsResponse:
  """
  Get all visits for a specific date for the current doctor.
  If no date is provided, returns today's visits.
  """
  target_date = date_param or date.today()
  rows = visits_service.list_by_doctor_and_date(current_doctor.doctor_id, target_date)
  
  visits = [_map_visit_with_patient(row) for row in rows]
  
  return TodayVisitsResponse(
    date=target_date,
    count=len(visits),
    visits=visits,
  )


@router.patch(
  "/{visit_id}",
  response_model=VisitResponse,
  status_code=status.HTTP_200_OK,
)
async def update_visit(
  visit_id: str,
  payload: VisitUpdateRequest,
  current_doctor: CurrentDoctor,
) -> VisitResponse:
  """Update visit details (date, time, notes, medications)."""
  update_values = payload.dict(exclude_unset=True)
  if not update_values:
    raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Empty update payload.")

  try:
    updated = visits_service.update_visit(visit_id, update_values)
  except ValueError as exc:
    raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc

  if not updated:
    raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Visit not found.")

  return VisitResponse(**updated)


@router.delete(
  "/{visit_id}",
  status_code=status.HTTP_204_NO_CONTENT,
)
async def delete_visit(
  visit_id: str,
  current_doctor: CurrentDoctor,
) -> None:
  """Delete a visit by ID."""
  # Check if visit exists
  visit = visits_service.get_visit_by_id(visit_id)
  if not visit:
    raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Visit not found.")
  
  # Delete the visit
  deleted = visits_service.delete_visit(visit_id)
  if not deleted:
    raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Visit not found.")


@router.patch(
  "/{visit_id}/status",
  response_model=VisitWithPatientResponse,
  status_code=status.HTTP_200_OK,
)
async def update_visit_status(
  visit_id: str,
  payload: VisitStatusUpdateRequest,
  current_doctor: CurrentDoctor,
) -> VisitWithPatientResponse:
  """
  Update visit status with business logic:
  - in_progress/completed: only for today or past visits
  - rescheduled: creates a new visit on the new date
  - no_show: marks patient as no-show
  """
  # Get current visit
  visit = visits_service.get_visit_by_id(visit_id)
  if not visit:
    raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Visit not found.")
  
  # Validate: in_progress/completed only for today or past
  visit_date = visit.get("visit_date")
  if isinstance(visit_date, str):
    visit_date = date.fromisoformat(visit_date)
  
  today = date.today()
  
  if payload.status in ("in_progress", "completed"):
    if visit_date and visit_date > today:
      raise HTTPException(
        status_code=status.HTTP_400_BAD_REQUEST,
        detail=f"Cannot set status '{payload.status}' for future visits."
      )
  
  # Validate: rescheduled requires new date
  if payload.status == "rescheduled":
    if not payload.rescheduled_to:
      raise HTTPException(
        status_code=status.HTTP_400_BAD_REQUEST,
        detail="rescheduled_to date is required when status is 'rescheduled'."
      )
    
    # Parse rescheduled_time if provided
    rescheduled_time = None
    if payload.rescheduled_time:
      try:
        rescheduled_time = time.fromisoformat(payload.rescheduled_time)
      except ValueError:
        raise HTTPException(
          status_code=status.HTTP_400_BAD_REQUEST,
          detail="Invalid rescheduled_time format. Use HH:MM."
        )
    
    # Update current visit as rescheduled
    visits_service.update_visit_status(
      visit_id,
      status="rescheduled",
      note=payload.note,
      rescheduled_to=payload.rescheduled_to,
      rescheduled_time=rescheduled_time,
    )
    
    # Create new visit on the rescheduled date
    new_visit_payload = {
      "visit_date": payload.rescheduled_to.isoformat(),
      "visit_time": payload.rescheduled_time if payload.rescheduled_time else visit.get("visit_time"),
      "notes": visit.get("notes"),
      "medications": visit.get("medications"),
    }
    
    visits_service.create_visit(
      doctor_id=visit.get("doctor_id", current_doctor.doctor_id),
      patient_id=visit.get("patient_id"),
      payload=new_visit_payload,
    )
  else:
    # Simple status update
    visits_service.update_visit_status(
      visit_id,
      status=payload.status,
      note=payload.note,
    )
  
  # Return updated visit with patient data
  updated = visits_service.get_visit_with_patient(visit_id)
  if not updated:
    raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Visit not found after update.")
  
  return _map_visit_with_patient(updated)

File: frontend/src/api/visits.ts
/**
 * Visits API Client
 * 
 * Provides methods for fetching and managing patient visits.
 */

import { apiClient, buildAuthHeaders } from './client'
import { getAuthToken } from './auth'

// Visit status enum
export const VISIT_STATUSES = [
  { value: 'scheduled', label: 'Scheduled', color: 'blue' },
  { value: 'in_progress', label: 'In Progress', color: 'yellow' },
  { value: 'completed', label: 'Completed', color: 'green' },
  { value: 'no_show', label: 'No Show', color: 'red' },
  { value: 'rescheduled', label: 'Rescheduled', color: 'gray' },
] as const

export type VisitStatus = (typeof VISIT_STATUSES)[number]['value']

export type PatientSummary = {
  id: string
  firstName: string
  lastName: string
  phone?: string | null
  telegramUserId?: number | null
  telegramUsername?: string | null
  whatsappPhone?: string | null
}

export type Visit = {
  id: string
  doctorId?: string
  patientId: string
  visitDate: string
  visitTime?: string | null
  nextVisitDate?: string | null
  notes?: string | null
  medications?: string | null
  status: VisitStatus
  statusChangedAt?: string | null
  statusNote?: string | null
  rescheduledTo?: string | null
  rescheduledTime?: string | null
  reminderStatus?: string | null
  reminderSentAt?: string | null
  reminderChannel?: string | null
  createdAt?: string
  patient?: PatientSummary | null
}

export type TodayVisitsResponse = {
  date: string
  count: number
  visits: Visit[]
}

export type UpdateVisitStatusInput = {
  status: VisitStatus
  note?: string
  rescheduledTo?: string
  rescheduledTime?: string
}

// API response types (snake_case from backend)
type ApiPatientSummary = {
  id: string
  first_name: string
  last_name: string
  phone?: string | null
  telegram_user_id?: number | null
  telegram_username?: string | null
  whatsapp_phone?: string | null
}

type ApiVisit = {
  id: string
  doctor_id?: string | null
  patient_id: string
  visit_date: string
  visit_time?: string | null
  next_visit_date?: string | null
  notes?: string | null
  medications?: string | null
  status: string
  status_changed_at?: string | null
  status_note?: string | null
  rescheduled_to?: string | null
  rescheduled_time?: string | null
  reminder_status?: string | null
  reminder_sent_at?: string | null
  reminder_channel?: string | null
  created_at?: string
  patient?: ApiPatientSummary | null
}

type ApiTodayVisitsResponse = {
  date: string
  count: number
  visits: ApiVisit[]
}

// Mappers
const mapPatientSummary = (data: ApiPatientSummary): PatientSummary => ({
  id: data.id,
  firstName: data.first_name,
  lastName: data.last_name,
  phone: data.phone ?? undefined,
  telegramUserId: data.telegram_user_id ?? undefined,
  telegramUsername: data.telegram_username ?? undefined,
  whatsappPhone: data.whatsapp_phone ?? undefined,
})

const isValidStatus = (status: string): status is VisitStatus =>
  VISIT_STATUSES.some((s) => s.value === status)

const mapVisit = (data: ApiVisit): Visit => ({
  id: data.id,
  doctorId: data.doctor_id ?? undefined,
  patientId: data.patient_id,
  visitDate: data.visit_date,
  visitTime: data.visit_time ?? undefined,
  nextVisitDate: data.next_visit_date ?? undefined,
  notes: data.notes ?? undefined,
  medications: data.medications ?? undefined,
  status: isValidStatus(data.status) ? data.status : 'scheduled',
  statusChangedAt: data.status_changed_at ?? undefined,
  statusNote: data.status_note ?? undefined,
  rescheduledTo: data.rescheduled_to ?? undefined,
  rescheduledTime: data.rescheduled_time ?? undefined,
  reminderStatus: data.reminder_status ?? undefined,
  reminderSentAt: data.reminder_sent_at ?? undefined,
  reminderChannel: data.reminder_channel ?? undefined,
  createdAt: data.created_at,
  patient: data.patient ? mapPatientSummary(data.patient) : undefined,
})

// API methods
export const visitsApi = {
  /**
   * Get today's visits for the current doctor
   */
  async getTodayVisits(): Promise<TodayVisitsResponse> {
    const authToken = getAuthToken()
    const { data } = await apiClient.get<ApiTodayVisitsResponse>('/visits/today', {
      headers: buildAuthHeaders(authToken),
    })
    return {
      date: data.date,
      count: data.count,
      visits: data.visits.map(mapVisit),
    }
  },

  /**
   * Get visits for a specific date
   */
  async getVisitsByDate(date: string): Promise<TodayVisitsResponse> {
    const authToken = getAuthToken()
    const { data } = await apiClient.get<ApiTodayVisitsResponse>('/visits', {
      headers: buildAuthHeaders(authToken),
      params: { date },
    })
    return {
      date: data.date,
      count: data.count,
      visits: data.visits.map(mapVisit),
    }
  },

  /**
   * Update visit status
   */
  async updateStatus(visitId: string, input: UpdateVisitStatusInput): Promise<Visit> {
    const authToken = getAuthToken()
    const payload: Record<string, string | undefined> = {
      status: input.status,
    }
    
    if (input.note !== undefined) {
      payload.note = input.note
    }
    if (input.rescheduledTo !== undefined) {
      payload.rescheduled_to = input.rescheduledTo
    }
    if (input.rescheduledTime !== undefined) {
      payload.rescheduled_time = input.rescheduledTime
    }

    const { data } = await apiClient.patch<ApiVisit>(
      `/visits/${visitId}/status`,
      payload,
      { headers: buildAuthHeaders(authToken) }
    )
    return mapVisit(data)
  },

  /**
   * Delete a visit
   */
  async deleteVisit(visitId: string): Promise<void> {
    const authToken = getAuthToken()
    await apiClient.delete(`/visits/${visitId}`, {
      headers: buildAuthHeaders(authToken),
    })
  },
}

export default visitsApi

